server-timing/class-perflab-server-timing.php

```diff
@@ -0,0 +1,255 @@
+			_doing_it_wrong(
+				__METHOD__,
+				/* translators: %s: WordPress action name */
+				sprintf( __( 'The method must be called before or during the %s action.', 'performance-lab' ), 'perflab_server_timing_send_header' ),
+				''
+			);
+			return;
+		}
+
+		$args = wp_parse_args(
+			$args,
+			array(
+				'measure_callback' => null,
+				'access_cap'       => null,
+			)
+		);
+		if ( ! $args['measure_callback'] || ! is_callable( $args['measure_callback'] ) ) {
+			_doing_it_wrong(
+				__METHOD__,
+				/* translators: %s: PHP parameter name */
+				sprintf( __( 'The %s argument is required and must be a callable.', 'performance-lab' ), '$args["measure_callback"]' ),
+				''
+			);
+			return;
+		}
+		if ( ! $args['access_cap'] || ! is_string( $args['access_cap'] ) ) {
+			_doing_it_wrong(
+				__METHOD__,
+				/* translators: %s: PHP parameter name */
+				sprintf( __( 'The %s argument is required and must be a string.', 'performance-lab' ), '$args["access_cap"]' ),
+				''
+			);
+			return;
+		}
+
+		$this->registered_metrics[ $metric_slug ]      = new Perflab_Server_Timing_Metric( $metric_slug );
+		$this->registered_metrics_data[ $metric_slug ] = $args;
+
+		// If the current user has already been determined and they lack the necessary access,
+		// do not even attempt to calculate the metric.
+		if ( did_action( 'set_current_user' ) && ! current_user_can( $args['access_cap'] ) ) {
+			return;
+		}
+
+		// Otherwise, call the measuring callback and pass the metric instance to it.
+		call_user_func( $args['measure_callback'], $this->registered_metrics[ $metric_slug ] );
+	}
+
+	/**
+	 * Checks whether the given metric has been registered.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $metric_slug The metric slug.
+	 * @return bool True if registered, false otherwise.
+	 */
+	public function has_registered_metric( $metric_slug ) {
+		return isset( $this->registered_metrics[ $metric_slug ] ) && isset( $this->registered_metrics_data[ $metric_slug ] );
+	}
+
+	/**
+	 * Outputs the Server-Timing header.
+	 *
+	 * This method must be called before rendering the page.
+	 *
+	 * @since n.e.x.t
+	 */
+	public function send_header() {
+		if ( headers_sent() ) {
+			_doing_it_wrong(
+				__METHOD__,
+				__( 'The method must be called before headers have been sent.', 'performance-lab' ),
+				''
+			);
+			return;
+		}
+
+		/**
+		 * Fires right before the Server-Timing header is sent.
+		 *
+		 * This action is the last possible point to register a Server-Timing metric.
+		 *
+		 * @since n.e.x.t
+		 */
+		do_action( 'perflab_server_timing_send_header' );
+
+		$header_value = $this->get_header_value();
+		if ( ! $header_value ) {
+			return;
+		}
+
+		header( sprintf( 'Server-Timing: %s', $header_value ), false );
+	}
+
+	/**
+	 * Gets the value for the Server-Timing header.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return string The Server-Timing header value.
+	 */
+	public function get_header_value() {
```

Comment:
this gets all the header values, maybe rename `get_header_values` (plural)

Reply:
@adamsilverstein I would kindly disagree here; the function returns multiple _Server Timing_ values, but it only returns the one header value for the `Server-Timing` header. That's why I went with this name. LMKWYT.

Reply:
don't feel strongly about this. I see it is a single value, just feels slightly wrong because the header contains value**s**. One other idea, just `get_header`?

Reply:
@adamsilverstein Good call! Will rename.